meson.build

```
@@ -0,0 +1,376 @@
+            description: 'Define to 1 if the ' + f[0] + '() function is available.'
+        )
+        break
+    endif
+endforeach
+
+if cc.has_function('posix_advice', prefix : '#include <fcntl.h>')
+    config_h.set10('HAVE_POSIX_ADVICE', true,
+        description: 'Define to 1 if the posix_advice() function is available.'
+    )
+endif
+
+### Filters
+
+## encoders
+encoder_msg = ''
+decoder_msg = ''
+
+supported_filters = [
+    'lzma1',
+    'lzma2',
+    'delta',
+    'x86',
+    'powerpc',
+    'ia64',
+    'arm',
+    'armthumb',
+    'arm64',
+    'sparc',
+    'riscv',
+]
+
+simple_filters = [
+    'x86',
+    'powerpc',
+    'ia64',
+    'arm',
+    'armthumb',
+    'arm64',
+    'sparc',
+    'riscv',
+]
+
+# FIXME foreach i in array; set_variable(f'have_@i@', true) endforeach
+
+foreach f : supported_filters
+    if get_option('encoders').contains(f)
+        config_h.set10('HAVE_ENCODER_' + f.underscorify().to_upper(), true,
+            description: 'Define to 1 if ' + f + ' encoder is enabled.'
+        )
+        encoder_msg += f + ' '
+    endif
+    if get_option('decoders').contains(f)
+        config_h.set10('HAVE_DECODER_' + f.underscorify().to_upper(), true,
+            description: 'Define to 1 if ' + f + ' decoder is enabled.'
+        )
+        decoder_msg += f + ' '
+    endif
+endforeach
+
+have_encoder_simple_filters = false
+have_decoder_simple_filters = false
+foreach f : simple_filters
+    if get_option('encoders').contains(f)
+        have_encoder_simple_filters = true
+    endif
+    if get_option('decoders').contains(f)
+        have_decoder_simple_filters = true
+    endif
+endforeach
+
+if (get_option('encoders').contains('lzma2') and not get_option('encoders').contains('lzma1')) or (get_option('decoders').contains('lzma2') and not get_option('decoders').contains('lzma1'))
+    error('LZMA2 requires that LZMA1 is also enabled')
+endif
+
+if get_option('encoders').length() > 0
+    config_h.set10('HAVE_ENCODERS', true,
+        description: 'Define to 1 if at least one encoder has been enabled.'
+    )
+endif
+
+if get_option('decoders').length() > 0
+    config_h.set10('HAVE_DECODERS', true,
+        description: 'Define to 1 if at least one decoder has been enabled.'
+    )
+endif
+
+### Match finders
+
+match_finder_msg = ''
+
+match_finders = [
+    'hc3',
+    'hc4',
+    'bt2',
+    'bt3',
+    'bt4'
+]
+
+if get_option('encoders').contains('lzma2') == true or get_option('encoders').contains('lzma1') == true
+    if get_option('match-finders').length() == 0
+        error('at least one match finder is required for an LZ-based encoder')
+    endif
+
+    foreach m : match_finders
+        if get_option('match-finders').contains(m) == true
+            config_h.set10('HAVE_MF_' + m.underscorify().to_upper(), true,
+                description: 'Define to 1 to enable ' + m + ' match finder.'
+            )
+            match_finder_msg += m + ' '
+        endif
+    endforeach
```

```suggestion
    foreach m : get_option('match-finders')
        config_h.set10('HAVE_MF_' + m.underscorify().to_upper(), true,
            description: 'Define to 1 to enable ' + m + ' match finder.'
        )
        match_finder_msg += m + ' '
    endforeach
```

Then you can get rid of the duplicated `match_finders` above, and keep it in one place: `meson_options.txt`. Meson will never let you set a value to this option that's not in the allowed options list.

---

done in 7056bbc2f62cb4c6e045ec93bcaca0faa9451dde